enhancements/monitoring/alert-overrides.md

```diff
@@ -0,0 +1,331 @@
+platform alerting rules.
+
+## Motivation
+
+OpenShift is an opinionated platform, and the default alerting rules have of
+course been carefully crafted to match what we think the majority of users need.
+Nevertheless, users have repeatedly asked for the ability to modify alerting
+rules to adjust thresholds, or to simply add labels and annotations to aide in
+routing alert notifications, as well as a supported method of adding additional
+alerting rules.
+
+Users currently have the ability to add additional rules from a technical
+standpoint -- by simply creating new `PrometheusRule` objects -- but this is
+[explicitly unsupported][1] at the moment.  The reason being that these
+resources are handled directly by the prometheus-operator, making it difficult
+to perform OpenShift specific defaulting and validation.  A concrete example
+being that there is no way to prevent users from creating new recording rules
+that may overwhelm the platform Prometheus instance.
+
+We've also encountered bugs and unexpected interactions that cause alerts to
+fire erroneously.  A supported method of patching platform alerting rules
+provides a more flexible method of applying temporary fixes in such cases.
+
+The goal of this proposal is to enable the addition of custom alerting rules,
+and limited ability to override existing platform alerting rules via a resource
+under the direct control of the cluster-monitoring-operator.
+
+### Goals
+
+- Give users a single, supported, and documented place to add additional rules,
+  in a way that allows the cluster-monitoring-operator to provide its own
+  defaulting, validation, etc.
+- Allow users to apply overrides to alerting rules shipped by the platform.
+- Allow users to drop unwanted platform alerts.
+- User-supplied overrides are merged with platform rules -- updates to fields
+  not modified by users continue to get updates.
+- Resources shipped by operators remain unmodified.  No action needed from
+  OpenShift developers.
+
+### Non-Goals
+
+- Only alerting rules are available for modification.  Not recording rules or
+  other monitoring configuration.
+
+## Proposal
+
+The cluster-monitoring-operator will be extended with a new controller that
+watches an instance of a new resource type with the following structure:
+
+```yaml
+apiVersion: v1alpha1
+kind: alertoverrides.monitoring.openshift.io
+metadata:
+  name: alert-overrides
+  namespace: openshift-monitoring
+spec:
+  # List of overrides for alerts in platform namespaces.
+  overrides:
+  - selector: # required
+      alert: "KubeAPIErrorBudgetBurn" # required
+      matchLabels: # optional, but multiple matches is an error reported in status.
+        severity: "critical"
+        long: "1h"
+    action: "patch"
+    labels:
+      patched: "true"
+
+  - selector:
+      alert: "KubeAPIErrorBudgetBurn"
+      matchLabels: # not a unique match, this is an error.
+        severity: "critical"
+    action: "patch"
+    labels:
+      patched: "true"
```

Comment:
Afaiu any patched rule will have a automatic label add to it, so that we can identify its as a patched alert. If a patch directive attempts to overwrite this generated label, we'll run into issues where users can mess with the ability to discern whether an alert stems from a shipped rule or from an overwritten one. For example assume here the generated labelname is `patched` and a patch rule adds `patched: "false"`.
The proposal should specifically call out this (error) scenario.

Reply:
This is a good point, but having thought about it again, we add the label identifying the alert as patched to the newly generated rule, and we do it after applying the user-supplied overrides. So in practice, it's not possible to change it. 

Also, in the relabel config, we match the original alerts by testing for the absence of the label. We don't actually care what the label's value is. So I don't think this presents a problem in the current design.

Reply:
Added a note about the label being added after patches are applied.